feat: lightpush new protobuf proposal (#20)
waku-org/specs: standards/core/lightpush.md

---

> title: WAKU2-LIGHTPUSH

Reviewer: Small comment: I think we should start only using the "waku" term. I see is actually well used in line 23.
```suggestion
title: WAKU-LIGHTPUSH
```

  - Oskar Thorén <oskarth@titanproxy.com>
---

previous version: `/vac/waku/lightpush/2.0.0-beta1` [19/WAKU2-LIGHTPUSH](https://rfc.vac.dev/waku/standards/core/19/lightpush)

---
**Protocol identifier**: `/vac/waku/lightpush/3.0.0`

## Motivation and Goals

Light nodes with short connection windows and limited bandwidth wish to push messages to other nodes in the Waku network to request message services.
A common use case is to request that the service node publish the message to an `11/WAKU2-RELAY` pubsub-topic.
Additionally, there is sometimes a need for confirmation that a message has been received "by the network"
(here, at least one node).

`WAKU-LIGHTPUSH` is a request/response protocol for this.

## Payloads

```protobuf
syntax = "proto3";

message LightPushRequest {
    string request_id = 1;
    uin32 request_type = 10;  // 10 Reserved for future `request_type`. Currently, RELAY is the only available service.
    optional string pubsub_topic = 20;
    WakuMessage message = 21;
}

message LightPushResponse {
    string request_id = 1;
    uint32 status_code = 10; // non zero in case of failure, see appendix
    optional string status_desc = 11;

>     uint32 relay_peer_count = 12; // number of peers, message is successfully relayed to 

Reviewer 2: Shouldn't this be optional? As this is specific to RELAY. In future if we add some other request_type, this field may not be required in the response.

---
slug: 19

> title: 19/WAKU2-LIGHTPUSH

Reviewer 3: No need for a number or slug. It can simply be titled WAKU2-LIGHTPUSH. Once implemented and tested, it could graduate to a Vac RFC and will get the number 19 then to indicate it replaces the current published spec.

Author: Ok, removed

> title: WAKU2-LIGHTPUSH
> name: Waku v2 Light Push

Reviewer: I think we can start avoiding using the "2" term and simply refer to Waku
```suggestion
title: WAKU-LIGHTPUSH
name: Waku Light Push
```

> status: draft

Reviewer 3: For a new protocol, remove the `status`
```suggestion
```

status: draft
editor: Zoltan Nagy <zoltan@status.im> 
contributors: 
  - Hanno Cornelius <hanno@status.im>
  - Daniel Kaiser <danielkaiser@status.im>
  - Oskar Thorén <oskarth@titanproxy.com>
---

previous version: `/vac/waku/lightpush/2.0.0-beta1` [19/WAKU2-LIGHTPUSH](https://rfc.vac.dev/waku/standards/core/19/lightpush)

---

> **Protocol identifier**: `/vac/waku/lightpush/2.0.0-beta2`

Reviewer 4: pet peeve of mine but I hate "beta" labels. Just increase the number! :stuck_out_tongue:

Reviewer 3: Agree. We can just call it 3.0.0

Author: Right, changed.

## Motivation and Goals

> Light nodes with short connection windows and limited bandwidth wish to publish messages into the Waku network.

Reviewer 3: Nitpick, but perhaps we can change the phrasing here in such a way that we do not commit to only one use case for lightpush, e.g.

```suggestion
Light nodes with short connection windows and limited bandwidth wish to push messages to other nodes in the Waku network to request message services.
A common use case is to request that service node to publish the message to a `11/WAKU2-RELAY` pubsub topic.
```

Additionally, there is sometimes a need for confirmation that a message has been received "by the network"
(here, at least one node).

> `WAKU2-LIGHTPUSH` is a request/response protocol for this.

Reviewer: ```suggestion
`WAKU-LIGHTPUSH` is a request/response protocol for this.
```

Author: Done

## Payloads

```protobuf
syntax = "proto3";

message LightPushRequest {
    enum RequestType { // Future extension point
        RELAY = 0;

>     }

Reviewer 4: In my mind we push a message to a node not to a protocol.

The only other protocol I can think of would be store.

:thinking: What would it mean to push to store instead of relay?

Author: @jm-clius  can explain I think, but mainly yes, push message to store but not relay or both and additionally the RFC refers to 

```forward the `LightPushRequest` via 19/LIGHTPUSH on a [WAKU2-DANDELION](https://github.com/waku-org/specs/blob/waku-RFC/standards/application/dandelion.md) stem```

Reviewer 5: Wouldn't pushing to Relay with the `ephemeral` field on be the same as pushing to both Relay and Store?

Author: How I understood this, to request just a store of the message but not to relay. Suppose we have a light client where using a lightpush service to directly store a message. @jm-clius is this what you meant?

Reviewer 3: > In my mind we push a message to a node not to a protocol.

Indeed, I think the point here is to allow for future differentiation of "message services" that you're requesting from the service node. In that sense it's not a list of protocols in the request type, but a list of possible services of which "relay publishing" may be one. Other possible services that we envision:
- store the message (without publishing to relay)
- dandelion hop (for anonymisation)
- add rln proof and publish (arguably a different service than just publishing)

However, @NagyZoltanPeter I think for this version of the protocol we can perhaps leave the enum undefined and just add a comment that tag/field number `10` is reserved for this use case? It would make more sense to add once we have more services.

Reviewer 4: > > In my mind we push a message to a node not to a protocol.
> 
> Indeed, I think the point here is to allow for future differentiation of "message services" that you're requesting from the service node. In that sense it's not a list of protocols in the request type, but a list of possible services of which "relay publishing" may be one. Other possible services that we envision:
> 
> * store the message (without publishing to relay)
> * dandelion hop (for anonymisation)
> * add rln proof and publish (arguably a different service than just publishing)

I re-read and I feel like it those use cases could be new protocols not extension. What's the benefit of doing it this way? The obvious downside is the added complexity of having multiple "type" of services in one protocol.

    
    string request_id = 1;
    RequestType kind = 10;  // default is RELAY, placeholder for future extensions

>     string pubsub_topic = 20;

Reviewer 4: pubsub topic would be the most agnostic but maybe we could use shard instead?

Reviewer 6: using shard can work, but guess we would need the cluster-id as well. since the pubsub topic encodes both.?

Author: I agree but IDK which wording would be the most straightforward to use. Do we have any in use?
Maybe `topic` as simple as is.

Author: WDYT? @alrevuelta, @jm-clius, @SionoiS, @Ivansete-status

Author: For a relay network it is still topic or pubsub_topic which routes the message, shard has meaning only under waku and not explained here well. I put a link there to direct interest.

>     uin32 request_type = 10;  // 10 Reserved for future `request_type`. Currently, RELAY is the only available service.

Reviewer 3: ```suggestion
    // 10 Reserved for future `request_type`. Currently, RELAY is the only available service.
```

I'd still leave this field out completely. In other words, add the comment but not the field.

>     uin32 request_type = 10;  // reserved for future request type selection, currently it is always RELAY (0)

Reviewer 3: Should be `uint32`. I would comment out this field completely for now though. Something like:
```suggestion
    // 10 Reserved for future `request_type`. Currently RELAY is only available service.
```

Author: Thanks.

>     string topic = 20;

Reviewer 3: ```suggestion
    optional string pubsub_topic = 20;
```
We should be explicit about this being a `pubsub_topic`. Note that making the field `optional` here doesn't mean that it is optional in the spec - it will simply allow the service node to verify that a `pubsub_topic` has been explicitly set by the client. Otherwise it will by default interpret any message withouth `pubsub_topic` as being directed at the empty string pubsub topic.

Author: Right, thank you.

    WakuMessage message = 21;
}

message LightPushResponse {
    enum Status {
        SUCCESS = 0;
        BAD_REQUEST = 400;
        NO_PEERS_TO_RELAY = 404;
        PAYLOAD_TOO_LARGE = 413;

>         UNSUPPORTED_TOPIC = 415;

Reviewer 6: would use contenttopic or pubsubtopic to avoid confusion.

Author: Er, what is the suggestion exactly? Maybe we can use different wording.
It refers to such requests having pubsub topic our service node is not subscribed to.
Of course it depends on the use case and configuration, what type of sharding is in use.

Reviewer 6: `UNSUPPORTED_TOPIC` -> `UNSUPPORTED_PUBSUBTOPIC`

Author: Renamed.

>         TOO_MANY_REQUESTS = 429;

Reviewer 6: Guess this maps to rln? Unsure how deep we could go but rln could lead to multiple errors. Some ideas:
* too many requests for epoch. assuming rlnv2 this means that the user has sent more messages than allowed within a given epoch.
* wrong epoch (either to early or to late)
* invalid root. the attached root of the proof does not match the window of acceptable roots.
* invalid proof. perhaps more generic to consider invalid proofs.

Author: Not exactly TOO_MANY_REQUESTS relates to rate limiting the service itself.
Bu I do agree, we can add more error codes that related to RLN. 
Also as agreed to move out the exact list of possible outcomes from the protocol definition it will be more easy to add any new if needed.

        INTERNAL_SERVER_ERROR = 500;
        SERVICE_UNAVAILABLE = 503;    

>     }

Reviewer 4: What if instead of defining the codes we keep this impl. detail? We can still specify that error codes should follow http standards.

Author: Hm, interesting idea not to have them in the communication interface description what outcome you might prepare for. For me it is more should be here than not.

Reviewer 3: Indeed, for new versions of protocols (and after a number of discussions), we've opted to leave `status_code` and `status_desc` generic `uint32` and `string` fields in order to define the possible error codes in the implementation. The reason being that we extend this list quite often and we'll soon get in a situation where the implementation and spec diverge. Of course, once we believe we have some stability in the services, we may publish a stand-beside spec with the list of supported error codes. See https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/12/filter.md#protobuf for how we're doing it in the new filter (and new store) protocols.

Author: Done. Added into a separate section as examples.

    string request_id = 1;
    Status status_code = 10;
    optional string status_desc = 11;

>     uint32 relay_peer_count = 12;

Reviewer 4: What if instead of sending back this number. We require the sender to specify the minimum # of peers sent to that they feel is appropriate?

Reviewer: IMO, the `lightPush` should not care about the `relay` connected peers. We can simply return an "LOW_NUMBER_OF_PEERS" error which will make the lightpush client to just try with another relay peer. The current number of relay peers can be returned back within the `status_desc` field

Reviewer 3: I think it makes sense for the service node to return the number of peers and the client to decide what to do with this information.

Author: I would close this as is here, returning exact peer count the message had been relayed to, gives the chance and flexibility to the client or SDK to decide its reactions. Other options would make the protocol just even fluid and harder to implement properly for other use cases.

Reviewer: @NagyZoltanPeter - I think it would be interesting to make it clear what the `relay_peer_count` means: whether it represents the actual number of connected nodes or the number of nodes that form the mesh. I'm happy with the decision but we might need to add a small comment somewhere to clarify that (even a comment next to this definition would be great :) )

( cc @jm-clius )

> Nodes that respond to `LightPushRequest` MUST either relay the encapsulated message via [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) protocol on the specified `pubsub_topic` or `shard` ([WAKU2-RELAY-SHARDING](https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md)) depending on the network configuration.

Reviewer 3: I would avoid using `shard` on this layer. The use of "either" here doesn't make sense as we don't provide an alternative.

Author: I reformatted this section.

> Nodes that respond to `LightPushRequest` MUST either relay the encapsulated message via [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) protocol on the specified `pubsub_topic`. Depending on the network configuration, the user may not need to provide `pubsub_topic` ([WAKU2-RELAY-SHARDING](https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md)).

Reviewer 3: ```suggestion
Nodes that respond to `LightPushRequest` SHOULD either relay the encapsulated message via [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) protocol on the specified `pubsub_topic`
or perform another requested service.
Services beyond [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) are as yet underdefined.
Depending on the network configuration, the lightpush client may not need to provide `pubsub_topic` ([WAKU2-RELAY-SHARDING](https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md)).
```
Please use https://sembr.org/

}
```

### Message Relaying

Nodes that respond to `LightPushRequest` MUST either
relay the encapsulated message via [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) protocol on the specified `pubsub_topic`,

> or forward the `LightPushRequest` via 19/LIGHTPUSH on a [WAKU2-DANDELION](https://github.com/waku-org/specs/blob/waku-RFC/standards/application/dandelion.md) stem.

Reviewer 4: Is this still relevant?

Reviewer 3: I would remove dandelion reference here (afaik it's not implemented) and just rephrase that by default the service provided is RELAY (without committing to this being the only way we'll process messages in future).

Author: Dandelion ref is removed.

> | NO_PEERS_TO_RELAY | 404 | Service node has no relay peers  |

Reviewer: A possible suggestion because the `NO_PEERS_TO_RELAY` value has a connotation of zero peers.
```suggestion
| NOT_ENOUGH_PEERS_TO_RELAY | 404 | Service node has no relay peers  |
```

Reviewer 3: I'd suggest sticking to HTTP codes as much as possible. A `404` usually implies an addressed resource was "not found" - a mistake the client needs to fix. Wouldn't "no peers to relay" match more closely to the "service unavailable" `503`? That is, you requested a service, but it is temporarily unavailable on the server-side due to lack of peers to relay to. We could in theory also use some of the unused HTTP codes to have more specific error cases.
Generally server-side errors should be in the `500` range.
Client-side errors (e.g. addressing a missing resource) should be in the `400` range.

See e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

Reviewer 7: Agree with @jm-clius and I think we should have: 
- SUCCESS - 200;
- BAD_REQUEST - to me is questionable, do we even need it? 
- UNSUPPORTED_TOPIC - maybe `421`?
- SERVICE_UNAVAILABLE overlaps with NO_PEERS_TO_RELAY - seems to me too;

Also, we need to figure out how to apply `418`! I always dreamt of doing so. 

cc @NagyZoltanPeter

Author: @weboko: I'm not sure that having non-zero success helps here, but I will do it to get it confirm with STORE's status/error codes.
418? :-D maybe for BAD_REQUEST, that's the closest, haha.

Author: Oh, and yes, BAD_REQUEST is anything if we cannot decode the request's payoad. It can be due to missused (wrong) protobuf definition.

Author: @weboko: For `SERVICE_UNAVAILABLE overlaps with NO_PEERS_TO_RELAY` question. I'm not pretty sure. There can be different reason why we don't have peers.
I'm thinking of: SERVICE_UNAVAILABLE - after startup we did not get connected to any peers yet, or we are disconnected for some reason. While NO_PEERS_TO_RELAY is rather related to the used pubsub_topic/shard we do not have connected peers that supports that. WDYT? I would keep both in this form.

If they are unable to do so for some reason, they SHOULD return an error code in `LightPushResponse`.

## Security Considerations

Since this can introduce an amplification factor, it is RECOMMENDED for the node relaying to the rest of the network to take extra precautions.
This can be done by rate limiting via [17/WAKU2-RLN-RELAY](https://rfc.vac.dev/waku/standards/core/17/rln-relay).

> Note that the above is currently not fully implemented.

Reviewer 4: It is now :smile:

Author: I rephrased the section.

> Once the relay is successful, the `relay_peer_count` will indicate the number of peers that the node has managed to relay the message to. It's important to note that this number may vary depending on the node subscriptions and support for the requested pubsub_topic. The client can use this information to either consider the relay as successful or take further action, such as switching to a lightpush service peer with better connectivity.

Reviewer 3: Semantic breaks

> | UNSUPPORTED_TOPIC | 415 | Requested push on pubsub_topic or shard is not possible as the service node is not subscibed to. |

Author: If that can be known than it is not a leak of info, isn't it? But can be valuable information for lightpush-client / SDK to decide to whether use or not use a service peer.
I'm also thinking of future planned dynamic bandwidth management feature where a node may decide to unsubscribe from some shards in order to lower resource needs (based on config).

## Future work

> - Add support attaching RLN proof for the message requested to be relayed.

Reviewer 4: I guess to be 100% future proof we should add this functionality.

Author: it is currently on going by Akhil.

> The field `relay_peer_count` may not present or has the value zero in case of error or in other future use cases, where no relay is involved.

Reviewer 3: ```suggestion
The field `relay_peer_count` may not be present or has the value zero in case of error or in other future use cases, where no relay is involved.
```

> Therefore Waku2 applies or will apply:

Reviewer: ```suggestion
Therefore Waku applies or will apply:
```

Author: Done

---
title: WAKU-LIGHTPUSH
name: Waku Light Push
editor: Zoltan Nagy <zoltan@status.im> 
contributors: 
  - Hanno Cornelius <hanno@status.im>
  - Daniel Kaiser <danielkaiser@status.im>
  - Oskar Thorén <oskarth@titanproxy.com>
---

previous version: `/vac/waku/lightpush/2.0.0-beta1` [19/WAKU2-LIGHTPUSH](https://rfc.vac.dev/waku/standards/core/19/lightpush)

---
**Protocol identifier**: `/vac/waku/lightpush/3.0.0`

## Motivation and Goals

> Light nodes with short connection windows and limited bandwidth wish to push messages to other nodes in the Waku network to request message services.<br>

Reviewer 8: instead of `<br>` (`<br />` seems to be the correct tag), just put an empty new line next.

A common use case is to request that the service node publish the message to an `11/WAKU2-RELAY` pubsub-topic.
Additionally, there is sometimes a need for confirmation that a message has been received "by the network"
(here, at least one node).

> `WAKU-LIGHTPUSH` is a request/response protocol for this.

Reviewer 8: Are we officially moving from `WAKU2` @jm-clius ?

Reviewer 3: Since this spec is not numbered yet (a.k.a. it's not a Vac RFC), the title is not critical.
For this to be promoted to Vac RFC, I suggest we keep using `WAKU2` in the official title (also for consistency with other protocols, including the new Store). We can drop the "v2" from the name though - this is simply Waku Lightpush.

The title will be:
`XX/WAKU2-LIGHTPUSH`

where XX is the Vac-assigned RFC number.

## Payloads

```protobuf
syntax = "proto3";

message LightPushRequest {
    string request_id = 1;
    // 10 Reserved for future `request_type`. Currently, RELAY is the only available service.
    optional string pubsub_topic = 20;
    WakuMessage message = 21;
}

message LightPushResponse {
    string request_id = 1;

>     uint32 status_code = 10; // non zero in case of failure, see appendix

Reviewer 8: > // non zero in case of failure, see appendix

Yes, `200` is described for success.

    optional string status_desc = 11;
    optional uint32 relay_peer_count = 12; // number of peers, message is successfully relayed to 
}
```

### Message Relaying

Nodes that respond to `LightPushRequest` SHOULD
- either relay the encapsulated message via [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) protocol on the specified `pubsub_topic`<br>
- or perform another requested service.

>   `Services beyond [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) are as yet underdefined.`

Reviewer 8: > `Services beyond [11/WAKU2-RELAY](https://rfc.vac.dev/waku/standards/core/11/relay) are as yet underdefined.`

simplify: "are yet to be defined"

Depending on the network configuration, the lightpush client may not need to provide `pubsub_topic` ([WAKU2-RELAY-SHARDING](https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md)).<br>

If the node is unable to perform the request for some reason, it SHOULD return an error code in `LightPushResponse`.<br>

> Once the relay is successful, the `relay_peer_count` will indicate the number of peers that the node has managed to relay the message to. It's important to note that this number may vary depending on the node subscriptions and support for the requested pubsub_topic. The client can use this information to either consider the relay as successful or take further action, such as switching to a lightpush service peer with better connectivity.<br>

Reviewer 8: sembr

>The field `relay_peer_count` may not be present or has the value zero in case of error or in other future use cases, where no relay is involved.

### Examples of possible error codes

| Result | Code | Note |
|--------|------|------|
| SUCCESS  | 200 | Successfull push, response's relay_peer_count holds the number of peers the message is pushed.    |
| BAD_REQUEST | 400   | Wrong request payload.    |
| PAYLOAD_TOO_LARGE | 413 | Message exceeds certain size limit, it can depend on network configuration, see status_desc for details.  |
| UNSUPPORTED_PUBSUB_TOPIC | 421 | Requested push on pubsub_topic is not possible as the service node does not support it. |
| TOO_MANY_REQUESTS | 429 | DOS protection prevented this request as the current request exceeds the configured request rate. |
| INTERNAL_SERVER_ERROR  | 500 | status_desc holds explanation of the error.  |
| NO_PEERS_TO_RELAY | 503 | Lightpush service is not available as the node has no relay peers. |

> The list of error codes is not complete and can be extended in the future.

> ## Security Considerations

Reviewer 8: I would add that the content of the response node does not provide guarantees that the messages was relayed and that the requesting node should mitigate appropriately.

Since this can introduce an amplification factor, it is RECOMMENDED for the node relaying to the rest of the network to take extra precautions.
Therefore Waku applies or will apply:
- DOS protection through request rate limitation on the service itself.
- message rate limiting via [17/WAKU2-RLN-RELAY](https://rfc.vac.dev/waku/standards/core/17/rln-relay), applied via network membership subscription.

> These features are under development. 

## Future work

- Add support attaching RLN proof for the message requested to be relayed.
- Add support for other request types.

> - Incentivization of the service

Reviewer 8: Definition of strategy to improve reliable usage of this service
